pilot/pkg/config/kube/gateway/controller.go

```diff
@@ -54,83 +53,280 @@ var log = istiolog.RegisterScope("gateway", "gateway-api controller")
+		DomainSuffix:    options.DomainSuffix,
+		Services:        inputs.Services,
+		Namespaces:      inputs.Namespaces,
+		ServiceEntries:  inputs.ServiceEntries,
+		internalContext: c.gatewayContext,
+	}
+	tcpRoutes := TCPRouteCollection(
+		inputs.TCPRoutes,
+		routeInputs,
+		opts,
+	)
+	registerStatus(c, tcpRoutes.Status)
+	tlsRoutes := TLSRouteCollection(
+		inputs.TLSRoutes,
+		routeInputs,
+		opts,
+	)
+	registerStatus(c, tlsRoutes.Status)
+	httpRoutes := HTTPRouteCollection(
+		inputs.HTTPRoutes,
+		routeInputs,
+		opts,
+	)
+	registerStatus(c, httpRoutes.Status)
+	grpcRoutes := GRPCRouteCollection(
+		inputs.GRPCRoutes,
+		routeInputs,
+		opts,
+	)
+	registerStatus(c, grpcRoutes.Status)
+
+	RouteAttachments := krt.JoinCollection([]krt.Collection[RouteAttachment]{
+		tcpRoutes.RouteAttachments,
+		tlsRoutes.RouteAttachments,
+		httpRoutes.RouteAttachments,
+		grpcRoutes.RouteAttachments,
+	}, opts.WithName("RouteAttachments")...)
+	RouteAttachmentsIndex := krt.NewIndex(RouteAttachments, func(o RouteAttachment) []types.NamespacedName {
+		return []types.NamespacedName{o.To}
 	})
 
-	if credsController != nil {
-		credsController.AddSecretHandler(gatewayController.secretEvent)
+	GatewayFinalStatus := FinalGatewayStatusCollection(GatewaysStatus, RouteAttachments, RouteAttachmentsIndex, opts)
+	registerStatus(c, GatewayFinalStatus)
+
+	VirtualServices := krt.JoinCollection([]krt.Collection[*config.Config]{
+		tcpRoutes.VirtualServices,
+		tlsRoutes.VirtualServices,
+		httpRoutes.VirtualServices,
+		grpcRoutes.VirtualServices,
+	}, opts.WithName("DerivedVirtualServices")...)
+
+	outputs := Outputs{
+		ReferenceGrants: ReferenceGrants,
+		Gateways:        Gateways,
+		VirtualServices: VirtualServices,
+	}
+	c.outputs = outputs
+
+	handlers = append(handlers,
+		outputs.VirtualServices.RegisterBatch(pushXds(xdsUpdater,
+			func(t *config.Config) model.ConfigKey {
+				return model.ConfigKey{
+					Kind:      kind.VirtualService,
+					Name:      t.Name,
+					Namespace: t.Namespace,
+				}
+			}), false),
+		outputs.Gateways.RegisterBatch(pushXds(xdsUpdater,
+			func(t Gateway) model.ConfigKey {
+				return model.ConfigKey{
+					Kind:      kind.Gateway,
+					Name:      t.Name,
+					Namespace: t.Namespace,
+				}
+			}), false),
+		outputs.ReferenceGrants.collection.RegisterBatch(pushXds(xdsUpdater,
+			func(t ReferenceGrant) model.ConfigKey {
+				return model.ConfigKey{
+					Kind:      kind.KubernetesGateway,
+					Name:      t.Source.Name,
+					Namespace: t.Source.Namespace,
+				}
+			}), false))
+	c.handlers = handlers
+
+	return c
+}
+
+// buildClient is a small wrapper to build a krt collection based on a delayed informer.
+func buildClient[I controllers.ComparableObject](
+	c *Controller,
+	kc kube.Client,
+	res schema.GroupVersionResource,
+	opts krt.OptionsBuilder,
+	name string,
+) krt.Collection[I] {
+	filter := kclient.Filter{
+		ObjectFilter: kubetypes.ComposeFilters(kc.ObjectFilter(), c.inRevision),
```

Comment:
I can confirm in the test, but how does this address namespace labeled revisions?